### PR TITLE
Extend and fine-tune OIDC error handling

### PR DESCRIPTION
**Pull request by moticless**
#### What does this PR do?
Extend and fine tune error handling of OIDC. Taking care to have proper log of the error and avoid from information disclosure by the response.

#### What ticket does this PR close?
Connected to #876

#### Where should the reviewer start?
authenticate_controller.rb
 
#### How should this be manually tested?
run request with expired token. Run request with invalid token. etc.

#### Has the Version and Changelog been updated?

> Can we make a blog post, video, or animated GIF of this?

> Has this change been documented (Readme, docs, etc.)?

> Does the knowledge base need an update?

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -5,6 +5,16 @@ class ApplicationController < ActionController::API
   include ::ActionView::Layouts
 
   class Unauthorized < RuntimeError
+
```
Comment (orenbm) [inline app/controllers/application_controller.rb]: please remove blank line

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -5,6 +5,16 @@ class ApplicationController < ActionController::API
...
+  class GatewayTimeout < RuntimeError
+  end
+
+  class BadGateway < RuntimeError
+  end
+
+  class BadRequest < RuntimeError
   end
 
   class BadRequest < RuntimeError
```
Comment (orenbm) [inline app/controllers/application_controller.rb]: is `BadRequest` defined twice in this file? it could have happen in the merge

File: app/domain/authentication/authn_oidc/authenticate_id_token/authenticate.rb [hunk 3]
```diff
@@ -72,7 +72,7 @@ def conjur_username
           id_token_username_field = oidc_secrets["id-token-user-property"]
 
           conjur_username = id_token_attributes[id_token_username_field]
-          raise IdTokenFieldNotFound, id_token_username_field unless conjur_username.present?
+          raise IdTokenFieldNotFound, id_token_username_field unless conjur_username
```
Comment (orenbm) [inline app/domain/authentication/authn_oidc/authenticate_id_token/authenticate.rb]: why was this changed? i had the `present?` to verify that not only the field exists but that it has a value. is this not needed?

File: app/domain/authentication/errors.rb [hunk 4]
```diff
@@ -22,9 +22,34 @@ module Authentication
...
+    ProviderDiscoveryTimeout = ::Util::ErrorClass.new(
+      "OIDC provider failed with timeout error (provider_uri='{0})'"
+    )
+
+    ProviderDiscoveryFailed = ::Util::ErrorClass.new(
+      "OIDC provider discovery failed (provider_uri='{0})'"
+    )
+
+    ProviderRetrieveCertificateFailed = ::Util::ErrorClass.new(
+      "Failed retrieve certificate from OIDC provider (provider_uri='{0})'"
```
Comment (orenbm) [inline app/domain/authentication/errors.rb]: Failed to retrieve certificate from the OIDC provider

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -185,7 +198,32 @@ def bad_request e
...
-    head :unauthorized
+    if e.message == Authentication::AuthnOidc::IdTokenExpired.new.to_s
+      render json: {
+        error: {
+          code: :unauthorized,
+          message: e.message
+        }
+      }, status: :unauthorized
+    else
+      logger.debug "#{e}\n#{e.backtrace.join "\n"}"
```
Comment (orenbm) [inline app/controllers/application_controller.rb]: this is redundant. we already log in the beginning of the method

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -185,7 +198,32 @@ def bad_request e
 
   def unauthorized e
     logger.debug "#{e}\n#{e.backtrace.join "\n"}"
-    head :unauthorized
+    if e.message == Authentication::AuthnOidc::IdTokenExpired.new.to_s
```
Comment (orenbm) [inline app/controllers/application_controller.rb]: this is very specific. what if we decide that also for another scenario we want to return the error message too? we'll need to update both `authenticate_controller.rb` and this file. please find a way that doesn't include the message check.

a possible solution is to check that the message is not "ApplicationController::Unauthorized" which is the default value (and thus if we want some exception to return the message to the user we raise it with message in `authenticate_controller.rb` as we do for `Authentication::AuthnOidc::IdTokenExpired`.

please check if there is a neater solution. maybe enhance this class in `application_controller.rb`
```
  class Unauthorized < RuntimeError
  end
```

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -27,6 +37,9 @@ class RecordExists < Exceptions::RecordExists
   rescue_from Exceptions::Forbidden, with: :forbidden
   rescue_from BadRequest, with: :bad_request
   rescue_from Unauthorized, with: :unauthorized
+  rescue_from GatewayTimeout, with: :gateway_timeout
+  rescue_from BadGateway, with: :bad_gateway
+  rescue_from BadRequest, with: :bad_request
```
Comment (orenbm) [inline app/controllers/application_controller.rb]: this is already defined in row 38

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -185,7 +198,32 @@ def bad_request e
...
+    logger.debug "#{e}\n#{e.backtrace.join "\n"}"
+    head :gateway_timeout
+  end
+
+  def bad_gateway e
+    logger.debug "#{e}\n#{e.backtrace.join "\n"}"
+    head :bad_gateway
+  end
+
+  def bad_request e
```
Comment (orenbm) [inline app/controllers/application_controller.rb]: this method is already defined in row 193

**Comment by orenbm** [comment]
@moticless did you add the log messages to the confluence table for review?

**Comment by nessiLahav** [comment]
Please add some Debug info logs to the authenticate flow

File: app/domain/authentication/authn_oidc/authenticate_id_token/authenticate.rb (hunk 3 above)
Comment (jtuttle) [reply app/domain/authentication/authn_oidc/authenticate_id_token/authenticate.rb]: I think this was correct with `present?`. Also interested to know why it changed.

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb [hunk 9]
```diff
@@ -41,12 +42,24 @@ def decoded_attributes
...
             @id_token_jwt,
-            get_cert(@provider_uri)
+            get_cert
           )
+        rescue => e
+          raise IdTokenInvalidFormat, e.to_s
+        end
+
+        def discovered_provider
+          @discovered_provider ||= OpenIDConnect::Discovery::Provider::Config.discover!(@provider_uri)
```
Comment (jtuttle) [inline app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: The `discover!` method being called here should not have a `!`. In Ruby, the `!` on a method name signals that it's a mutator method that modifies the object that calls it. I'm guessing that's not happening here, though correct me if I'm wrong!

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb [hunk 10]
```diff
@@ -11,7 +11,8 @@ module AuthenticateIdToken
       ) do
 
         def call
-          decode_id_token
+          discovered_provider
+          decoded_id_token
```
Comment (jtuttle) [inline app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: Are you calling these methods early to make the `call` method look like an index of what's happening? I wouldn't bother doing so in this situation. It's forcing you to memoize an object that you could just be returning (`@discovered_provider`) and reads a bit strangely since these two methods are essentially 'getters'.  You could reword the methods to be a bit more action oriented like `discover_provider` and `decode_id_token` perhaps. Otherwise, I would just remove them and let them happen as helpers in `verify_id_token`.

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb (hunk 9 above)
Comment (jtuttle) [reply app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: Also if you  make the change I suggested above to the `call` method, I don't think you need to memoize `@ discovered_provider` and can just return it from the method for use in `get_cert`.

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb [hunk 11]
```diff
@@ -41,12 +42,24 @@ def decoded_attributes
...
+          raise ProviderDiscoveryTimeout, @provider_uri
+        rescue => e
+          raise ProviderDiscoveryFailed, @provider_uri
         end
 
-        def get_cert(provider_uri)
-          OpenIDConnect::Discovery::Provider::Config.discover!(provider_uri).jwks
+        def get_cert
+          discovered_provider.jwks
+        rescue => e
```
Comment (jtuttle) [inline app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: Any chance we get a useful error here that might tell us what went wrong? If so, we should try to preserve it.

File: app/domain/authentication/errors.rb [hunk 12]
```diff
@@ -22,9 +22,34 @@ module Authentication
 
   module AuthnOidc
 
+    ProviderDiscoveryTimeout = ::Util::ErrorClass.new(
+      "OIDC provider failed with timeout error (provider_uri='{0})'"
```
Comment (jtuttle) [inline app/domain/authentication/errors.rb]: Should be "provider **discovery** failed"

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb (hunk 9 above)
Comment (orenbm) [reply app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: the `discover!` method is in the 3rd party

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb (hunk 10 above)
Comment (orenbm) [reply app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: yes, that was the idea. would you rather have something like this?

```
def decode_id_token
  decoded_id_token
end
```
?

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb (hunk 9 above)
Comment (jtuttle) [reply app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: Oh weird, okay!

File: app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb (hunk 10 above)
Comment (jtuttle) [reply app/domain/authentication/authn_oidc/authenticate_id_token/decode_and_verify_id_token.rb]: Hmm good question. @jonahx what's the "command class" philosophy here? It seems odd to be calling helper methods early just to produce a readable index. Should they be removed? Or is there a bigger restructure that needs to happen here to create a `call` method with meaningful line items?

File: app/domain/authentication/authn_oidc/authenticate_id_token/authenticate.rb (hunk 3 above)
Comment (moticless) [reply app/domain/authentication/authn_oidc/authenticate_id_token/authenticate.rb]: You are right. The only issue is that the error not fully reflect the error. I will modify the name and the log to IdTokenFieldNotFoundOrEmpty
